src/protocols/virtual_pointer.rs

```diff
@@ -0,0 +1,554 @@
+use std::collections::HashSet;
+use std::sync::Mutex;
+
+use smithay::backend::input::{
+    AbsolutePositionEvent, Axis, AxisRelativeDirection, AxisSource, ButtonState, Device,
+    DeviceCapability, Event, InputBackend, PointerAxisEvent, PointerButtonEvent,
+    PointerMotionAbsoluteEvent, PointerMotionEvent, UnusedEvent,
+};
+use smithay::input::pointer::AxisFrame;
+use smithay::reexports::wayland_protocols_wlr;
+use smithay::reexports::wayland_server::protocol::wl_output::WlOutput;
+use smithay::reexports::wayland_server::protocol::wl_pointer;
+use smithay::reexports::wayland_server::protocol::wl_seat::WlSeat;
+use smithay::reexports::wayland_server::{
+    Client, DataInit, Dispatch, DisplayHandle, GlobalDispatch, New, Resource,
+};
+use wayland_backend::protocol::WEnum;
+use wayland_protocols_wlr::virtual_pointer::v1::server::{
+    zwlr_virtual_pointer_manager_v1, zwlr_virtual_pointer_v1,
+};
+use zwlr_virtual_pointer_manager_v1::ZwlrVirtualPointerManagerV1;
+use zwlr_virtual_pointer_v1::ZwlrVirtualPointerV1;
+
+const VERSION: u32 = 2;
+
+pub struct VirtualPointerManagerState {
+    virtual_pointers: HashSet<ZwlrVirtualPointerV1>,
+}
+
+pub struct VirtualPointerManagerGlobalData {
+    filter: Box<dyn for<'c> Fn(&'c Client) -> bool + Send + Sync>,
+}
+
+pub struct VirtualPointerInputBackend;
+
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+pub struct VirtualPointer {
+    pointer: ZwlrVirtualPointerV1,
+}
+
+#[derive(Debug)]
+pub struct VirtualPointerUserData {
+    seat: Option<WlSeat>,
+    output: Option<WlOutput>,
```

Comment:
What is this supposed to do?

Reply:
Uh,, it stores the parameters the virtual pointer was created with? I don't think i used it for anything (and looking at references, i just noticed that i kept in `#[allow(dead_code)]` for the getters); but this is like, "generic protocol code" that could (should?) be part of Smithay or copied to other compositors, so i figured it would be weird to not keep it in.

Something we *should* be doing, though, is properly handling absolute pointer events when `output` is `Some(_)`, which i don't think i did. This would just happen to work by chance on single-monitor layouts, but i think on multi-monitor it would look janky without special care taken (as is done for tablets' and touchscreens' `map-to-output`)

Reply:
> i just noticed that i kept in #[allow(dead_code)] for the getters

fixed this now by making them pub instead, which also silences the warnings. since i never ended up using them, ig i forgot that i should've made them pub. lol.

Reply:
I guess it's for absolute motion, yeah. How hard would that be to implement?

Reply:
From a glance at the input handling, it looks like it would be nontrivial to make this track the correct output. We could fairly easily just bypass the existing input handling; but I think it'd be best to have some trait we implement on each input backend to allow us to query further device information about each event, like which output this absolute position event logically refers to, if any. This should apply equally to all devices, and would probably imply the ability to configure individual devices in the config, and virtual pointers are just one kind of device that is configured via wlr-virtual-pointer instead.

This is... maybe too much just for this PR? should I open another one? Alternatively, maybe Smithay should be able to handle providing an output for an absolute motion event? At the very least, that Wayland backend I've been working on does require the ability to specify which output a motion event belongs to, if we're ever gonna hope to see it use multiple outputs mapped as multiple toplevels. That could make more sense to implement in Smithay than in niri.

Likewise, it makes a lot of sense for individual keyboard devices to carry their xkb states and keymaps, for e.g. virtual keyboard, which currently Smithay just sneaks past the compositor and directly into its surface tree, but ideally they should go through the normal input backend system, and propagate their keymaps properly. The same goes for my Wayland backend, which can have multiple keyboards in different outer seats with different layouts. A lot of that could/should make its way into Smithay, while at the same time it wouldn't really be too difficult to implement such extension traits in niri.

Reply:
There's already a map-to-output system in niri for tablets and touchscreens, and there's a downcast to libinput event to figure out what device the event comes from. Maybe you could similarly downcast to a virtual pointer event, and then get the output from it?

Reply:
> There's already a map-to-output system in niri for tablets and touchscreens

Yeah, I know. But these are applied unilaterally to all touch and tablet position events, based on the event kind; it's not very useful to apply that to virtual pointers since there isn't a single output they all should map to. What I'm suggesting is we replace this system with one that can associate the mapped output with the device or the event data.

Reply:
Yeah but if you downcast the absolute motion pointer event to its virtual absolute motion pointer event type, then you can grab the right output from that event. Should be simple enough until proper systems are in place in Smithay or wherever.

Reply:
https://github.com/YaLTeR/niri/pull/630#issuecomment-2495646090